src/eval.ml

```diff
@@ -187,34 +196,55 @@ and tree_walk : dtree -> stack -> (term * stack) option = fun tree stk ->
                 env.(slot) <- TE_Some(b)
               else
                 let b = lift vars.(pos) in
-                let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+                let xs = Array.map (fun e -> IntMap.find e id_vars) xs in
                 env.(slot) <- TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs b))
         in
         List.iter fn env_builder;
         (* Actually perform the action. *)
         Some(Bindlib.msubst act env, stk)
-    | Cond({ ok ; cond ; fail })                          ->
+    | Cond({ok; cond; fail})                              ->
         let next =
           match cond with
           | CondNL(i, j) ->
               if eq_modulo vars.(i) vars.(j) then ok else fail
           | CondFV(xs,i) ->
-              let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+              let allowed =
+                (* Term variables allowed in the term. *)
+                let _, dummy = IntMap.choose id_vars in
+                Array.make (Array.length xs) dummy in
+              let lightened_map =
+                (* *Lightened* map containing variables from the context
+                   without the variables in the constraint (the [xs]). This
+                   avoids checking occurrence of variables that have been
+                   bound in the term (via [bind_mvar]).
+                   Aditionally, the [allowed] array is filled imperatively. *)
+                let c = Pervasives.ref 0 in
```

Comment:
What is [c] for? It is used nowhere.

Reply:
Ah indeed there is an error: it ought to be used to fill the [allowed] array, instead of the [i]. It's strange that it works with this.